Add get and set methods to Capabilities

PunitLodha opened the pull request:
Fixes #421
I had a few doubts,
- What should be done, in the get method, if there's no such capability. If we return an error then I'll have to Box the errors. Or is there any other way to do it?
- In the set method, what should be the label for the capability, if its a custom_capability

PunitLodha committed:
add get and set methods to Capabilities

jplatte commented:
For get: I think we should just return `Result<Option<_>>`.

For set: it should have a capability string parameter before the json value to determine which capability to set.

PunitLodha committed:
update get and set methods
change return type of get method to Result<Option<_>>
add a capability label parameter to set method

jplatte commented on ruma-client-api/src/r0/capabilities/get_capabilities.rs:
```diff
@@ -72,14 +72,45 @@ pub struct Capabilities {
+    /// Returns value of the required capability
+    pub fn get(&self, capability: &str) -> Result<Option<Cow<'_, JsonValue>>, Error> {
+        let value = match capability {
+            "m.change_password" => Some(Cow::Owned(to_value(self.change_password.clone())?)),
```
```suggestion
            "m.change_password" => Some(Cow::Owned(to_value(&self.change_password)?)),
```

jplatte replied:
Same below.

jplatte commented on ruma-client-api/src/r0/capabilities/get_capabilities.rs:
```diff
@@ -5,8 +5,8 @@ use ruma_api::ruma_api;
-use serde_json::Value as JsonValue;
-use std::collections::BTreeMap;
+use serde_json::{from_value, to_value, Error, Value as JsonValue};
+use std::{borrow::Cow, collections::BTreeMap};
```
Please move this into its own import group. (see `CONTRIBUTING.md`)

jplatte commented on ruma-client-api/src/r0/capabilities/get_capabilities.rs:
```diff
@@ -72,14 +72,45 @@ pub struct Capabilities {
         Default::default()
     }
+
+    /// Returns value of the required capability
```
```suggestion
    /// Returns value of the given capability. 
```

jplatte commented on ruma-client-api/src/r0/capabilities/get_capabilities.rs:
```diff
@@ -72,14 +72,45 @@ pub struct Capabilities {
+    pub fn set(&mut self, capability_label: &str, capability: JsonValue) -> Result<(), Error> {
+        match capability_label {
+            "m.change_password" => {
+                let change_password_capability: ChangePasswordCapability = from_value(capability)?;
```
You can inline this, the type annotation shouldn't actually be needed.

jplatte reviewed (commented):
Looks good overall, thanks! Some nits below.

jplatte commented:
Also, the PR title still says `[WIP]` (GitHub allows you to mark PRs as draft instead, by the way).

PunitLodha committed:
refactor

jplatte commented:
FWIW: I made some adjustments after merging this. Turns out we don't need `Result` in `get` at all, since the only way we can have an `Err` is serialization of one of our own types failing. Since we derive `Serialize` and don't use any crazy attributes, I'm pretty sure the only way serialization can fail is through a bug in `serde` or `serde_json`. I also added some more docs.
